### PR TITLE
Drivers: Uart: Add hardware flow control support for STM32

### PR DESCRIPTION
**Pull request by Benichou34**
Add option for enable hardware flow control support for STM32.
This option can be individually activated for each uart.

### CODE REVIEW

File: drivers/pinmux/stm32/pinmux_stm32f0.h [hunk 1]
```diff
@@ -10,91 +10,361 @@
...
  */
+/* Port A */
+#define STM32F0_PINMUX_FUNC_PA0_USART1_CTS \
+	(STM32_PINMUX_ALT_FUNC_1 | STM32_PUPDR_NO_PULL)
+#define STM32F0_PINMUX_FUNC_PA0_USART2_CTS \
+	(STM32_PINMUX_ALT_FUNC_1 | STM32_PUPDR_NO_PULL)
+#define STM32F0_PINMUX_FUNC_PA0_USART4_TX \
+	(STM32_PINMUX_ALT_FUNC_4 | STM32_PUSHPULL_PULLUP)
 
-#define STM32F0_PINMUX_FUNC_PB6_USART1_TX \
```
Comment (erwango) [inline drivers/pinmux/stm32/pinmux_stm32f0.h]: It's ok to reorganize the file, but please do it in a separate commit, this making it hard to review.

File: drivers/serial/Kconfig.stm32 [hunk 2]
```diff
@@ -25,6 +25,13 @@ config UART_1
 	  Enable support for USART1 port in the driver.
 	  Say y here if you want to use USART1 device.
 
+config UART_1_HW_FC
```
Comment (erwango) [inline drivers/serial/Kconfig.stm32]: Please check how we could  introduce dtscts support as a device tree property (uart-has-rtscts) instead of Kconfig.

File: drivers/serial/uart_stm32.c [hunk 3]
```diff
@@ -685,7 +689,8 @@ static const struct uart_stm32_config uart_stm32_cfg_##name = {		\
 };									\
 									\
 static struct uart_stm32_data uart_stm32_data_##name = {		\
-	.baud_rate = DT_UART_STM32_##name##_BAUD_RATE			\
+	.baud_rate = DT_UART_STM32_##name##_BAUD_RATE,			\
+	.hw_fc = STM32_##name##_HW_FC \
```
Comment (erwango) [inline drivers/serial/uart_stm32.c]: You could have used CONFIG_##name##_HW_FC directly here and spare the lines in each device declaration below

File: drivers/serial/uart_stm32.c [hunk 4]
```diff
@@ -630,6 +630,10 @@ static int uart_stm32_init(struct device *dev)
 	/* Set the default baudrate */
 	uart_stm32_set_baudrate(dev, data->baud_rate);
 
+	if (data->hw_fc) {
```
Comment (erwango) [inline drivers/serial/uart_stm32.c]: A comment would be nice.

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/pinmux/stm32/pinmux_stm32f0.h (hunk 1 above)
Comment (Benichou34) [reply drivers/pinmux/stm32/pinmux_stm32f0.h]: I have created a new pull request for this.

**Comment by Benichou34** [comment]
Hardware flow control can be individually activated for each uart in the device tree (hardware-fc-rtscts).

File: soc/arm/st_stm32/stm32f0/dts_fixup.h [hunk 5]
```diff
@@ -71,6 +72,34 @@
 #define DT_USART_2_IRQ				DT_ST_STM32_USART_40004400_IRQ_0
 #define DT_UART_STM32_USART_2_CLOCK_BITS	DT_ST_STM32_USART_40004400_CLOCK_BITS
 #define DT_UART_STM32_USART_2_CLOCK_BUS	DT_ST_STM32_USART_40004400_CLOCK_BUS
+#define DT_UART_STM32_USART_2_HARDWARE_FC_RTSCTS	DT_ST_STM32_USART_40004400_HARDWARE_FC_RTSCTS
+
+#define DT_UART_STM32_USART_3_BASE_ADDRESS   DT_ST_STM32_USART_40004800_BASE_ADDRESS
```
Comment (erwango) [inline soc/arm/st_stm32/stm32f0/dts_fixup.h]: New instances addition should be done in a separate commit (ideally a separate PR)

File: dts/bindings/serial/uart.yaml [hunk 6]
```diff
@@ -24,6 +24,11 @@ properties:
       category: required
       description: Initial baud rate setting for UART
       generation: define
+    hardware-fc-rtscts:
+      type: boolean
+      category: optional
+      description: Initial hardware flow control setting for UART
```
Comment (erwango) [inline dts/bindings/serial/uart.yaml]: IIRC, behavior of optional booleans is as follows:
If available in the node, generated define will be set to true
If not available, generated define will be set to true
So, I'll put the description as follows: "Hardware flow control enabled at boot time".

File: drivers/serial/uart_stm32.c [hunk 7]
```diff
@@ -685,7 +690,8 @@ static const struct uart_stm32_config uart_stm32_cfg_##name = {		\
 };									\
 									\
 static struct uart_stm32_data uart_stm32_data_##name = {		\
-	.baud_rate = DT_UART_STM32_##name##_BAUD_RATE			\
+	.baud_rate = DT_UART_STM32_##name##_BAUD_RATE,			\
+	.hw_fc = DT_UART_STM32_##name##_HARDWARE_FC_RTSCTS \
```
Comment (erwango) [inline drivers/serial/uart_stm32.c]: Is that field strictly required? We're able to store/load info from IP registers so ideally we don't need to use RAM bits to store the info.

File: dts/bindings/serial/uart.yaml [hunk 8]
```diff
@@ -24,6 +24,11 @@ properties:
       category: required
       description: Initial baud rate setting for UART
       generation: define
+    hardware-fc-rtscts:
```
Comment (erwango) [inline dts/bindings/serial/uart.yaml]: Name seems ok to me but I'd like @galak to comment on this.

**Review by erwango — CHANGES_REQUESTED** [review]
This looks much better using dts!
Please also update uart_stm32_configure/uart_stm32_configure_get apis.

Last, please split your commits into smaller logical entities (yaml, drivers, pins..).

File: dts/bindings/serial/uart.yaml (hunk 8 above)
Comment (erwango) [reply dts/bindings/serial/uart.yaml]: @Benichou34 , seems that a `hw-flow-control` is already available for the same purpose, can you check?

Comment (Benichou34) [reply dts/bindings/serial/uart.yaml]: I used the "hw-flow-control" as dts binding keyword.

File: soc/arm/st_stm32/stm32f0/dts_fixup.h (hunk 5 above)
Comment (Benichou34) [reply soc/arm/st_stm32/stm32f0/dts_fixup.h]: I separated the modified files into several commits.

**Comment by erwango** [comment]
Thanks for update!
Meanwhile, another PR has been proposed for the same feature (#13163) by @GeorgeCGV. Can you get along and see which one to keep?

**Comment by GeorgeCGV** [comment]
@erwango @Benichou34 
I thought this PR stalled as there was nothing going on for some time. In this PR we miss configuration set, get implementations; and pinmux for F4, and L4.
Apart from that, they are similar.
All of the above can be added here or it can be merged separately, I don't mind.
